core/drivers/imx_i2c.c

```diff
@@ -0,0 +1,397 @@
+	{ 3072,	0x1E }, { 3840,	0x1F }
+};
+
+static void i2c_set_prescaler(uint8_t bid, uint32_t rate)
+{
+	struct ifdr_pair *p = ifdr_table;
+	struct ifdr_pair *q = p + ARRAY_SIZE(ifdr_table) - 1;
+	uint32_t div = (I2C_CLK_RATE + rate - 1) / rate;
+
+	if (div < p->divider)
+		q = p;
+	else if (div > q->divider)
+		p = q;
+
+	while (p != q) {
+		if (div <= p->divider)
+			break;
+		p++;
+	}
+
+	i2c_io_write8(bid, IFDR, p->prescaler);
+}
+
+static void enable_i2c_clk(uint8_t bid)
+{
+	struct imx_i2c_clk *c = &i2c_clk;
+
+	io_write32(c->base.va + CCM_CCGRx_SET(c->i2c[bid]),
+		   CCM_CCGRx_ALWAYS_ON(0));
+}
+
+static void i2c_set_bus_speed(uint8_t bid, int speed)
+{
+	enable_i2c_clk(bid);
+	i2c_set_prescaler(bid, speed);
+}
+
+static TEE_Result i2c_sync_bus(uint8_t bid, bool (*match)(uint32_t),
+			       uint32_t *status)
+{
+	uint64_t tref = timeout_init_us(100000);
+	uint32_t sr = 0;
+
+	while (!timeout_elapsed(tref)) {
+		sr = i2c_io_read8(bid, I2SR);
+		if (sr & I2SR_IAL) {
+			EMSG("bus arbitration lost");
+			i2c_io_write8(bid, I2SR, sr & ~I2SR_IAL);
+			return TEE_ERROR_COMMUNICATION;
+		}
+		if ((*match)(sr)) {
+			if (status)
+				*status = sr;
+			return TEE_SUCCESS;
+		}
+	}
+
+	return TEE_ERROR_BUSY;
+}
+
+static TEE_Result i2c_idle_bus(uint8_t bid)
+{
+	uint8_t tmp = i2c_io_read8(bid, I2CR) & ~I2CR_MSTA;
+	TEE_Result ret = TEE_SUCCESS;
+
+	i2c_io_write8(bid, I2CR, tmp);
+	ret = i2c_sync_bus(bid, &bus_is_idle, NULL);
+	i2c_io_write8(bid, I2SR, 0);
+
+	return ret;
+}
+
+static TEE_Result i2c_write_byte(uint8_t bid, uint8_t byte)
+{
+	TEE_Result ret = TEE_SUCCESS;
+	uint32_t status = 0;
+
+	i2c_io_write8(bid, I2DR, byte);
+	ret = i2c_sync_bus(bid, &isr_active, &status);
+	i2c_io_write8(bid, I2SR, 0);
+
+	if (ret == TEE_SUCCESS && (status & I2SR_RX_NO_AK))
+		return TEE_ERROR_BAD_STATE;
+
+	return TEE_SUCCESS;
+}
+
+static TEE_Result i2c_read_byte(uint8_t bid, uint8_t *p)
+{
+	TEE_Result ret = TEE_SUCCESS;
+
+	*p = i2c_io_read8(bid, I2DR);
+	ret = i2c_sync_bus(bid, &isr_active, NULL);
+	i2c_io_write8(bid, I2SR, 0);
+
+	return ret;
+}
+
+static TEE_Result i2c_write_data(uint8_t bid, const uint8_t *buf, int len)
+{
+	TEE_Result ret = TEE_SUCCESS;
+	uint32_t tmp = 0;
+
+	if (!len)
+		return TEE_SUCCESS;
+
+	tmp = i2c_io_read8(bid, I2CR) | I2CR_MTX | I2CR_TX_NO_AK;
+	i2c_io_write8(bid, I2CR, tmp);
+
+	while (len--) {
+		ret = i2c_write_byte(bid, *buf++);
+		if (ret != TEE_SUCCESS)
+			return ret;
+	}
+
+	return ret;
+}
+
+static TEE_Result i2c_read_data(uint8_t bid, uint8_t *buf, int len)
+{
+	TEE_Result ret = TEE_SUCCESS;
+	uint8_t dummy = 0;
+	uint32_t tmp = 0;
+
+	if (!len)
+		return TEE_SUCCESS;
+
+	tmp = i2c_io_read8(bid, I2CR) & ~I2CR_MTX;
+	tmp = (len == 1) ? tmp | I2CR_TX_NO_AK : tmp & ~I2CR_TX_NO_AK;
+	i2c_io_write8(bid, I2CR, tmp);
+	i2c_io_read8(bid, I2DR);
+
+	ret = i2c_read_byte(bid, &dummy);
+	if (ret != TEE_SUCCESS)
+		return ret;
+
+	/*
+	 * i.MX 8M Mini APRM Rev2 08/2019:
```

Comment:
Could you remove this line? It is not i.MX8M specific

Reply:
yes